### PR TITLE
[`bnb`] Fix bnb skip modules

### LINKED ISSUE #24037: BitsAndBytesConfig llm_int8_skip_modules does not work in the new version
### System Info

- `transformers` version: 4.30.0.dev0
- Platform: Linux-5.4.0-148-generic-x86_64-with-glibc2.17
- Python version: 3.8.16
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
from transformers import RobertaForSequenceClassification, BitsAndBytesConfig
quantization_config = BitsAndBytesConfig(load_in_8bit=True, llm_int8_skip_modules=['classifier'])
model = RobertaForSequenceClassification.from_pretrained('roberta-large-mnli', quantization_config=quantization_config)
```

### Expected behavior

The 'classifier' layer should be in Float16 but is actually loaded in 8bit. This is problematic because it drastically lower the performance of the model. It also make it impossible to train it (using peft library for example).

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/24037

https://github.com/huggingface/transformers/pull/23479 removed by mistake the logic introduced in https://github.com/huggingface/transformers/pull/21579 to deal with modules that are not needed to be converted

The PR also adds a nice test to make sure this will never happen again

### CODE REVIEW

File: tests/bitsandbytes/test_mixed_int8.py [hunk 1]
```diff
@@ -146,6 +146,17 @@ def test_linear_are_8bit(self):
...
 
+    def test_llm_skip(self):
+        r"""
+        A simple test to check if `llm_int8_skip_modules` works as expected
+        """
+        quantization_config = BitsAndBytesConfig(load_in_8bit=True, llm_int8_skip_modules=["classifier"])
+        seq_classification_model = AutoModelForSequenceClassification.from_pretrained(
+            "roberta-large-mnli", quantization_config=quantization_config
+        )
+        self.assertTrue(isinstance(seq_classification_model.classifier.dense, nn.Linear))
```
Comment (amyeroberts) [inline tests/bitsandbytes/test_mixed_int8.py:157]: Just for my own understanding (not a comment to address), here we're checking the layers of the classifier are `nn.Linear`. In `test_linear_are_8bit`, we check that the layers are `nn.Linear` too and that their dtype is `torch.int8` (I didn't know this was possible!). Are we certain that this means these layers are loaded in correctly? Do we need a dtype check on the weights?

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

File: tests/bitsandbytes/test_mixed_int8.py [hunk 2]
```diff
@@ -146,6 +146,17 @@ def test_linear_are_8bit(self):
...
+        self.assertTrue(isinstance(seq_classification_model.classifier.out_proj, nn.Linear))
+
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply tests/bitsandbytes/test_mixed_int8.py:159]: Awesome yes agreed! Will add that now

File: tests/bitsandbytes/test_mixed_int8.py (hunk 1 above)
Comment (younesbelkada) [reply tests/bitsandbytes/test_mixed_int8.py:157]: You are right, we also need a dtype check on the weights! `Linear8bitLt` has `nn.Linear` as a super class. Adding new tests!

**Pull request merged**